libnfs4acl/nfs4_ace_to_json.c

```diff
@@ -0,0 +1,357 @@
+ * SUCH DAMAGE.
+ */
+
+#include <stdio.h>
+#include <ctype.h>
+#include <assert.h>
+#include <jansson.h>
+#include "nfs4_json.h"
+#include "libacl_nfs4.h"
+
+static int
+who_to_json(json_t *_parent, struct nfs4_ace *entry, bool numeric)
+{
+	int error;
+	char *tag = NULL;
+	char who_str[NFS4_MAX_PRINCIPALSIZE] = {0};
+	int who_id = -1;
+
+	switch (entry->whotype) {
+	case NFS4_ACL_WHO_NAMED:
+		tag = NFS4_IS_GROUP(entry->flag) ? "group" : "user";
+		if (numeric) {
+			error = acl_nfs4_get_who(entry, &who_id, NULL, 0);
+			if (error) {
+				return (error);
+			}
+		}
+		else {
+			error = acl_nfs4_get_who(entry, &who_id, &who_str,
+						 NFS4_MAX_PRINCIPALSIZE);
+			if (error) {
+				return (error);
+			}
+		}
+		break;
+	case NFS4_ACL_WHO_OWNER:
+		tag = "owner@";
+		break;
+	case NFS4_ACL_WHO_GROUP:
+		tag = "group@";
+		break;
+	case NFS4_ACL_WHO_EVERYONE:
+		tag = "everyone@";
+		break;
+	default:
+		return (-1);
+
+	}
+	error = json_object_set_new(_parent, "tag", json_string(tag));
+	if (error) {
+		return (error);
+	}
+	if (!numeric && entry->whotype == NFS4_ACL_WHO_NAMED) {
+		error = json_object_set_new(_parent, "who", json_string(who_str));
+		if (error) {
+			return (error);
+		}
+	}
+	error = json_object_set_new(_parent, "id", json_integer(who_id));
+	if (error) {
+		return (error);
+	}
+
+	return (0);
+}
+
+static int
+perms_to_json(json_t *_parent, nfs4_acl_perm_t access_mask, bool verbose)
+{
+	int error, i;
+	json_t *basic = NULL, *perms = NULL;
+	perms = json_object();
+	if (perms == NULL) {
+		return (-1);
+	}
+	if (!verbose) {
+		if (access_mask == (NFS4_ACE_FULL_SET)) {
+			basic = json_string("FULL_CONTROL");
+			if (basic == NULL) {
+				return (-1);
+			}
+		}
+		else if (access_mask == (NFS4_ACE_MODIFY_SET)) {
+			basic = json_string("MODIFY");
+			if (basic == NULL) {
+				return (-1);
+			}
+		}
+		else if (access_mask == (NFS4_ACE_READ_SET |
+					 NFS4_ACE_EXECUTE)) {
+			basic = json_string("READ");
+			if (basic == NULL) {
+				return (-1);
+			}
+		}
+		else if (access_mask == (NFS4_ACE_EXECUTE |
+					 NFS4_ACE_READ_ATTRIBUTES |
+					 NFS4_ACE_READ_NAMED_ATTRS |
+					 NFS4_ACE_READ_ACL)) {
+			basic = json_string("TRAVERSE");
+			if (basic == NULL) {
+				return (-1);
+			}
+		}
+		if (basic) {
+			error = json_object_set_new(perms, "BASIC", basic);
+			if (error) {
+				return (error);
+			}
+			goto done;
+		}
+
+	}
+	for (i = 0; i < ARRAY_SIZE(perms2txt); i++) {
+		error = json_object_set_new(
+		    perms,
+		    perms2txt[i].name,
+		    json_boolean(access_mask & perms2txt[i].perm ? true : false));
+
+		if (error) {
+			return (error);
+		}
+	}
+
+done:
+	error = json_object_set_new(_parent, "perms", perms);
+	return (error);
+}
+
+static int
+flags_to_json(json_t *_parent, nfs4_acl_flag_t flagset, bool verbose)
+{
+	int error, i;
+	json_t *basic = NULL, *flags = NULL;
+	flags = json_object();
+	if (flags == NULL) {
+		return (-1);
+	}
+	if (!verbose) {
+		if (flagset == (NFS4_ACE_DIRECTORY_INHERIT_ACE |
```

Comment:
Likewise on the switch/definitions (maybe enums even, so the compiler can check for all cases being handled).